src/fs.rs

```diff
@@ -395,6 +401,246 @@ impl Fs {
+                        Utc::now(),
+                        &mut fs.private_forest,
+                        &mut fs.store,
+                        &mut rand::thread_rng(),
+                    )
+                    .await?;
+                let file = file.clone();
+                DetachedWriter::new_private(path, file, fs.private_forest.clone())
+            }
+            PathSegments::Public(path) => DetachedWriter::new_public(path),
+        };
+        Ok(this)
+    }
+
+    pub fn new_public(path: Vec<String>) -> Self {
+        Self {
+            path,
+            inner: DetachedWriterInner::Public(None),
+        }
+    }
+    pub fn new_private(path: Vec<String>, file: PrivateFile, forest: Rc<PrivateForest>) -> Self {
+        Self {
+            path,
+            inner: DetachedWriterInner::Private { file, forest },
+        }
+    }
+    pub async fn write_all(
+        &mut self,
+        content: impl AsyncRead + Send + Unpin + 'static,
+        store: &mut store::Store,
+    ) -> anyhow::Result<()> {
+        match &mut self.inner {
+            DetachedWriterInner::Public(cid) => {
+                let res_cid = store
+                    .put_block_streaming(content, libipld::IpldCodec::Raw)
+                    .await?;
+                *cid = Some(res_cid);
+            }
+            DetachedWriterInner::Private { file, forest } => {
+                file.set_content(
+                    Utc::now(),
+                    content.compat(),
+                    forest,
+                    store,
+                    &mut rand::thread_rng(),
+                )
+                .await?;
+            }
+        }
+        Ok(())
+    }
+
+    pub async fn finalize(self, fs: &mut Fs) -> anyhow::Result<()> {
+        match self.inner {
+            DetachedWriterInner::Public(cid) => {
+                let cid = cid
+                    .ok_or_else(|| anyhow::anyhow!("Called finalize before write was finished"))?;
+                fs.public
+                    .write(&self.path, cid, Utc::now(), &fs.store)
+                    .await?;
+            }
+            DetachedWriterInner::Private { file, forest } => {
+                // Merge forests
+                let private_forest = Rc::make_mut(&mut fs.private_forest);
+                *private_forest = private_forest.merge(&forest, &mut fs.store).await?;
+                // Get a handle to the file in the upstream PrivateDirectory
+                let upstream_file = fs
+                    .private
+                    .open_file_mut(
+                        &self.path,
+                        true,
+                        Utc::now(),
+                        &mut fs.private_forest,
+                        &mut fs.store,
+                        &mut rand::thread_rng(),
+                    )
+                    .await?;
+                // Assign the written-to file to the file within the PrivateDirectory.
+                *upstream_file = file;
```

Comment:
Oh wow. Yeah I didn't even know this was possible but it makes sense :joy:

This should work, but will overwrite concurrent changes on the same file IIUC.

:man_shrug:

To do what I proposed initially I guess you'd need an async mutex (one that you can hold over await points).

Reply:
Concurrent changes to the same file from FUSE are already not possible, the `fuse::WriteHandle` checks that writes to a file are append-only and returns an error for non-sequential writes.

With an async mutex and without the `DetachedWriter',  I think I'd still have to hold a global lock on the `Fs` in a `WriteHandle` until completion of the file (because the `PrivateFile::set_content` takes mutable references) which would block all FUSE operations while a write is in progress.